src/base/memory_monitor_sr.cpp

```diff
@@ -281,61 +319,111 @@ namespace cubperf
 
 	    if (!comp_skip)
 	      {
-		comp_idx = add_component (info[cnt].comp_name);
+		comp_idx = add_component (info[idx].comp_name);
 	      }
-	    /* XXX: add error if comp_idx > MAX_COMP_IDX */
+	    assert (comp_idx < MAX_COMP_IDX);
 
-	    if (info[cnt].subcomp_name)
+	    if (info[idx].subcomp_name)
 	      {
-		subcomp_idx = m_component[comp_idx]->get_subcomp_index (info[cnt].subcomp_name);
+		subcomp_idx = m_component[comp_idx]->get_subcomp_index (info[idx].subcomp_name);
 		if (subcomp_idx != -1)
 		  {
 		    subcomp_skip = true;
 		  }
 
 		if (!subcomp_skip)
 		  {
-		    subcomp_idx = m_component[comp_idx]->add_subcomponent (info[cnt].subcomp_name);
+		    subcomp_idx = m_component[comp_idx]->add_subcomponent (info[idx].subcomp_name);
 		  }
-		/* XXX: add error if subcomp_idx > MAX_SUBCOMP_IDX */
+		assert (subcomp_idx < MAX_COMP_IDX);
 	      }
 	  }
+	else
+	  {
+	    assert (info[idx].subcomp_name == NULL);
+	  }
 
-	m_comp_idx_map.push_back (make_stat_idx_map (comp_idx, subcomp_idx));
+	m_comp_idx_map.push_back (make_comp_idx_map_val (comp_idx, subcomp_idx));
 
-	cnt++;
+	idx++;
       }
   }
 
   int mmon_module::aggregate_stats (const MMON_MODULE_INFO &info)
   {
-    return 0;
+    int error = NO_ERROR;
+    return error;
   }
 
-  void mmon_module::add_stat (uint64_t size, int comp_idx, int subcomp_idx, bool init, bool expand)
+  void mmon_module::add_stat (uint64_t size, int comp_idx, int subcomp_idx, bool allocation_at_init)
   {
     /* description of add_stat(..).
-     * 1) if init == true, add init_stat
+     * 1) if allocation_at_init == true, add init_stat
      * 2) then, add cur_stat
      * 3) compare with peak_stat
      * 4) if cur_stat(new) > peak_stat, update peak_stat
-     * 5) if expand == true, expand_count++
-     * 6) call component->add_stat(size, subcomp_idx, init, expand) */
-    return;
+     * 5) call component->add_stat(size, subcomp_idx, allocation_at_init) */
+
+    if (allocation_at_init)
+      {
+	m_stat.init_stat += size;
+      }
+
+    m_stat.cur_stat += size;
+
+    if (m_stat.cur_stat > m_stat.peak_stat)
+      {
+	m_stat.peak_stat = m_stat.cur_stat.load ();
+      }
+
+    /* If there is a component info */
+    if (comp_idx != mmon_module::MAX_COMP_IDX)
```

Comment:
Based on the comment in the memory_monitor::add_stat()
1. introduce a new function to get comp and subcomp index
    - for example
        ```
        int mmon_module::get_comp_and_subcomp_id (MMON_STAT_ID stat_id, int &comp_id, int &subcump_id)
        ```
    - with this, I expect to get rid of the MMON_GET_COMP_MAP_IDX_FROM_STAT_ID macro
2. move the condition check into the newly introduced function and if the index is decided to the unexpected range, then reture error code and handle it in this function